# 2023-07-13 -- Voltron Data Webinars -- Using Selectors for productive data wrangling

The goal of this webinar is to show how you can use selectors in Ibis to speed up your data wrangling.

We will demonstrate how to use selectors on a small (< 3 Mb) dataset both locally using DuckDB as the backend and in the cloud with Snowflake.
The data here is a weather dataset that we chose because it really shows how Ibis makes it easy to wrangle it for downstream applications like visiualization or machine learning but selectors can save you time and effort on any type of dataset.
.columns
## Setup

To get started, we'll import Ibis and read the dataset.

In [50]:
from ibis.interactive import * 

In [51]:
weather_raw = ibis.read_csv("weather_prediction_dataset.csv")

Let's inspect the dataset:

In [52]:
print(weather_raw.count())
print(len(weather_raw.columns))
weather_raw.columns

3654

165


['DATE',
 'MONTH',
 'BASEL_cloud_cover',
 'BASEL_humidity',
 'BASEL_pressure',
 'BASEL_global_radiation',
 'BASEL_precipitation',
 'BASEL_sunshine',
 'BASEL_temp_mean',
 'BASEL_temp_min',
 'BASEL_temp_max',
 'BUDAPEST_cloud_cover',
 'BUDAPEST_humidity',
 'BUDAPEST_pressure',
 'BUDAPEST_global_radiation',
 'BUDAPEST_precipitation',
 'BUDAPEST_sunshine',
 'BUDAPEST_temp_mean',
 'BUDAPEST_temp_max',
 'DE_BILT_cloud_cover',
 'DE_BILT_wind_speed',
 'DE_BILT_wind_gust',
 'DE_BILT_humidity',
 'DE_BILT_pressure',
 'DE_BILT_global_radiation',
 'DE_BILT_precipitation',
 'DE_BILT_sunshine',
 'DE_BILT_temp_mean',
 'DE_BILT_temp_min',
 'DE_BILT_temp_max',
 'DRESDEN_cloud_cover',
 'DRESDEN_wind_speed',
 'DRESDEN_wind_gust',
 'DRESDEN_humidity',
 'DRESDEN_global_radiation',
 'DRESDEN_precipitation',
 'DRESDEN_sunshine',
 'DRESDEN_temp_mean',
 'DRESDEN_temp_min',
 'DRESDEN_temp_max',
 'DUSSELDORF_cloud_cover',
 'DUSSELDORF_wind_speed',
 'DUSSELDORF_wind_gust',
 'DUSSELDORF_humidity',
 'DUSSELDORF_pres

It has 3654 rows, 165 columns. Each row represents the observation for a day (column `DATE`) and then the city (uppercase) followed by several weather features (wind, humidity, pressure, global radiation, precipitation, temperature (mean, min, and max)). There 18 European cities from 11 countries with dates from 2000 to 2010 represented in this dataset. You can find more information about this dataset here: https://zenodo.org/record/7525955

For the purpose of this webinar, we are going to select the data from the 3 French cities found in this dataset. Ibis makes this easy using selectors:

In [53]:
( 
    weather_raw
    .select(s.c("DATE", "MONTH"), s.startswith(("MONTELIMAR", "PERPIGNAN", "TOURS")))
)

┏━━━━━━━━━━┳━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┓
┃ DATE     ┃ MONTH ┃ MONTELIMAR_wind_speed ┃ MONTELIMAR_humidity ┃ MONTELIMAR_pressure ┃ MONTELIMAR_global_radiation ┃ MONTELIMAR_precipitation ┃ MONTELIMAR_temp_mean ┃ MONTELIMAR_temp_min ┃ MONTELIMAR_temp_max ┃ PERPIGNAN_wind_speed ┃ PERPIGNAN_humidity ┃ PERPIGNAN_pressure ┃ PERPIGNAN_global_radiation ┃ PERPIGNAN_precipitation ┃ PERPIGNAN_temp_mean ┃ PERPIGNAN_temp_min ┃ PERPIGNAN_temp_max ┃ TOURS_wind_speed ┃ TOURS_humidity ┃ TOURS_pressure ┃ TOURS_global_radiation ┃ TOURS_precipitation ┃ TOURS_temp_mean ┃ TOURS_temp_min ┃ TOURS_temp_max ┃
┡━━━━━━━━━━╇━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━┩
│ int64    │ int64 │ float64               │ float64             │ float64             │ float64                     │ float64                  │ float64              │ float64             │ float64             │ float64              │ float64            │ float64            │ float64                    │ float64                 │ float64             │ float64            │ float64            │ float64          │ float64        │ float64        │ float64                │ float64             │ float64         │ float64        │ float64        │
├──────────┼───────┼───────────────────────┼─────────────────────┼─────────────────────┼─────────────────────────────┼──────────────────────────┼──────────────────────┼─────────────────────┼─────────────────────┼──────────────────────┼────────────────────┼────────────────────┼────────────────────────────┼─────────────────────────┼─────────────────────┼────────────────────┼────────────────────┼──────────────────┼────────────────┼────────────────┼────────────────────────┼─────────────────────┼─────────────────┼────────────────┼────────────────┤
│ 20000101 │     1 │                   3.8 │                0.85 │              1.0269 │                        0.30 │                      0.0 │                  5.5 │                 2.5 │                 8.5 │                  4.4 │               0.71 │             1.0267 │                       0.60 │                    0.00 │                12.2 │               10.3 │               14.0 │              1.6 │           0.97 │         1.0275 │                   0.25 │                0.04 │             8.5 │            7.2 │            9.8 │
│ 20000102 │     1 │                   5.8 │                0.82 │              1.0287 │                        0.54 │                      0.0 │                  8.3 │                 6.8 │                 9.8 │                  2.9 │               0.67 │             1.0278 │                       0.96 │                    0.00 │                 9.8 │                5.1 │               14.6 │              2.0 │           0.99 │         1.0293 │                   0.17 │                0.16 │             7.9 │            6.6 │            9.2 │
│ 20000103 │     1 │                   0.4 │                0.92 │              1.0316 │                        0.53 │                      0.0 │            

In this format, the data isn't very easy to use. We want to have a column with the city names, and remove these names from the column labels. The `pivot_longer` and `pivot_wider` functions are going to help with this: 

In [54]:
(
    weather_raw
    .select(s.c("DATE", "MONTH"), s.startswith(("MONTELIMAR", "PERPIGNAN", "TOURS")))
    .pivot_longer(
        ~ s.c("DATE", "MONTH"),
        names_to = ["city", "variable"],
        names_pattern = r"(^[A-Z]+)_(.+)"
    )
    .pivot_wider(
        names_from = "variable",
        values_from = "value"
    )
)

┏━━━━━━━━━━┳━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┓
┃ DATE     ┃ MONTH ┃ city       ┃ wind_speed ┃ humidity ┃ pressure ┃ global_radiation ┃ precipitation ┃ temp_mean ┃ temp_min ┃ temp_max ┃
┡━━━━━━━━━━╇━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━┩
│ int64    │ int64 │ string     │ float64    │ float64  │ float64  │ float64          │ float64       │ float64   │ float64  │ float64  │
├──────────┼───────┼────────────┼────────────┼──────────┼──────────┼──────────────────┼───────────────┼───────────┼──────────┼──────────┤
│ 20030501 │     5 │ MONTELIMAR │        2.1 │     0.67 │   1.0208 │             2.32 │          0.02 │      14.3 │      6.9 │     21.7 │
│ 20030501 │     5 │ PERPIGNAN  │        3.0 │     0.67 │   1.0200 │             1.57 │          0.00 │      14.3 │      9.0 │     19.6 │
│ 20030501 │     5 │ TOURS      │        4.2 │     0.75 │   1.0178 │             1.84 │          0.00 │      13.4 │      9.0 │     17.7 │
│ 20030502 │     5 │ MONTELIMAR │        2.5 │     0.60 │   1.0174 │             2.24 │          0.00 │      17.5 │     10.4 │     24.6 │
│ 20030502 │     5 │ PERPIGNAN  │        4.5 │     0.69 │   1.0172 │             2.44 │          0.00 │      16.2 │     12.4 │     20.0 │
│ 20030502 │     5 │ TOURS      │        5.9 │     0.74 │   1.0139 │             2.03 │          0.06 │      13.9 │     10.4 │     17.4 │
│ 20030503 │     5 │ MONTELIMAR │        6.0 │     0.51 │   1.0214 │             3.14 │          0.00 │      17.3 │     13.3 │     21.3 │
│ 20030503 │     5 │ PERPIGNAN  │        3.2 │     0.40 │   1.0209 │             3.24 │          0.00 │      19.1 │     13.7 │     24.5 │
│ 20030503 │     5 │ TOURS      │        3.0 │     0.73 │   1.0220 │             2.36 │          0.00 │      12.8 │      7.1 │     18.5 │
│ 20030504 │     5 │ MONTELIMAR │        2.8 │     0.61 │   1.0188 │             2.84 │          0.00 │      17.4 │      7.1 │     27.7 │
│        … │     … │ …          │          … │        … │        … │                … │             … │         … │        … │        … │
└──────────┴───────┴────────────┴────────────┴──────────┴──────────┴──────────────────┴───────────────┴───────────┴──────────┴──────────┘

To give you an idea of what it would have taken to write the equivalent SQL code for this data wrangling operation, let's use the `ibis.show_sql` function on this previous statement:

In [55]:
ibis.show_sql(
    weather_raw
    .select(s.c("DATE", "MONTH"), s.startswith(("MONTELIMAR", "PERPIGNAN", "TOURS")))
    .pivot_longer(
        ~ s.c("DATE", "MONTH"),
        names_to = ["city", "variable"],
        names_pattern = r"(^[A-Z]+)_(.+)"
    )
    .pivot_wider(
        names_from = "variable",
        values_from = "value"
    )
)

WITH t0 AS (
  SELECT
    t3."DATE" AS "DATE",
    t3."MONTH" AS "MONTH",
    t3."MONTELIMAR_wind_speed" AS "MONTELIMAR_wind_speed",
    t3."MONTELIMAR_humidity" AS "MONTELIMAR_humidity",
    t3."MONTELIMAR_pressure" AS "MONTELIMAR_pressure",
    t3."MONTELIMAR_global_radiation" AS "MONTELIMAR_global_radiation",
    t3."MONTELIMAR_precipitation" AS "MONTELIMAR_precipitation",
    t3."MONTELIMAR_temp_mean" AS "MONTELIMAR_temp_mean",
    t3."MONTELIMAR_temp_min" AS "MONTELIMAR_temp_min",
    t3."MONTELIMAR_temp_max" AS "MONTELIMAR_temp_max",
    t3."PERPIGNAN_wind_speed" AS "PERPIGNAN_wind_speed",
    t3."PERPIGNAN_humidity" AS "PERPIGNAN_humidity",
    t3."PERPIGNAN_pressure" AS "PERPIGNAN_pressure",
    t3."PERPIGNAN_global_radiation" AS "PERPIGNAN_global_radiation",
    t3."PERPIGNAN_precipitation" AS "PERPIGNAN_precipitation",
    t3."PERPIGNAN_temp_mean" AS "PERPIGNAN_temp_mean",
    t3."PERPIGNAN_temp_min" AS "PERPIGNAN_temp_min",
    t3."PERPIGNAN_temp_max" AS "PERPIGNAN_temp_max"

There are two other things we need to fix:
- we need to standardize the column names: the first two columns are uppercase while the rest is lower case. Let's make them all lowercase
- the date is stored as an integer, let's convert it into a proper date

In [56]:
(
    weather_raw
    .select(s.c("DATE", "MONTH"), s.startswith(("MONTELIMAR", "PERPIGNAN", "TOURS")))
    .pivot_longer(
        ~ s.c("DATE", "MONTH"),
        names_to = ["city", "variable"],
        names_pattern = r"(^[A-Z]+)_(.+)"
    )
    .pivot_wider(
        names_from = "variable",
        values_from = "value"
    )
    .relabel(str.lower)
    .mutate(
        date = _.date.cast("str").to_timestamp("%Y%m%d").cast("date")
    )
)

┏━━━━━━━━━━━━┳━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┓
┃ date       ┃ month ┃ city       ┃ wind_speed ┃ humidity ┃ pressure ┃ global_radiation ┃ precipitation ┃ temp_mean ┃ temp_min ┃ temp_max ┃
┡━━━━━━━━━━━━╇━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━┩
│ date       │ int64 │ string     │ float64    │ float64  │ float64  │ float64          │ float64       │ float64   │ float64  │ float64  │
├────────────┼───────┼────────────┼────────────┼──────────┼──────────┼──────────────────┼───────────────┼───────────┼──────────┼──────────┤
│ 2000-01-01 │     1 │ MONTELIMAR │        3.8 │     0.85 │   1.0269 │             0.30 │          0.00 │       5.5 │      2.5 │      8.5 │
│ 2000-01-01 │     1 │ PERPIGNAN  │        4.4 │     0.71 │   1.0267 │             0.60 │          0.00 │      12.2 │     10.3 │     14.0 │
│ 2000-01-01 │     1 │ TOURS      │        1.6 │     0.97 │   1.0275 │             0.25 │          0.04 │       8.5 │      7.2 │      9.8 │
│ 2000-01-02 │     1 │ MONTELIMAR │        5.8 │     0.82 │   1.0287 │             0.54 │          0.00 │       8.3 │      6.8 │      9.8 │
│ 2000-01-02 │     1 │ PERPIGNAN  │        2.9 │     0.67 │   1.0278 │             0.96 │          0.00 │       9.8 │      5.1 │     14.6 │
│ 2000-01-02 │     1 │ TOURS      │        2.0 │     0.99 │   1.0293 │             0.17 │          0.16 │       7.9 │      6.6 │      9.2 │
│ 2000-01-03 │     1 │ MONTELIMAR │        0.4 │     0.92 │   1.0316 │             0.53 │          0.00 │       3.2 │     -2.1 │      8.5 │
│ 2000-01-03 │     1 │ PERPIGNAN  │        2.5 │     0.85 │   1.0288 │             0.93 │          0.00 │       8.6 │      4.1 │     13.2 │
│ 2000-01-03 │     1 │ TOURS      │        3.4 │     0.91 │   1.0267 │             0.27 │          0.00 │       8.1 │      6.6 │      9.6 │
│ 2000-01-04 │     1 │ MONTELIMAR │        1.1 │     0.85 │   1.0274 │             0.64 │          0.00 │       7.2 │      2.3 │     12.1 │
│ …          │     … │ …          │          … │        … │        … │                … │             … │         … │        … │        … │
└────────────┴───────┴────────────┴────────────┴──────────┴──────────┴──────────────────┴───────────────┴───────────┴──────────┴──────────┘

We are going to create a few columns that will allow us to do some interesting queries with this dataset.
1. We are going to create a year column
2. We are going to create a month column but as a short name for the month
3. We are going to create an amplitude column that will be the difference between the max and the min temperatures

In [57]:
(
    weather := weather_raw
    .select(s.c("DATE", "MONTH"), s.startswith(("MONTELIMAR", "PERPIGNAN", "TOURS")))
    .pivot_longer(
        ~ s.c("DATE", "MONTH"),
        names_to = ["city", "variable"],
        names_pattern = r"(^[A-Z]+)_(.+)"
    )
    .pivot_wider(
        names_from = "variable",
        values_from = "value"
    )
    .relabel(str.lower)
    .mutate(
        date = _.date.cast("str").to_timestamp("%Y%m%d").cast("date")
    )
    .mutate(
        year = _.date.year(),
        month_str = _.date.strftime('%b'),
        temp_amplitude = _.temp_max - _.temp_min
    )
)

┏━━━━━━━━━━━━┳━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┓
┃ date       ┃ month ┃ city       ┃ wind_speed ┃ humidity ┃ pressure ┃ global_radiation ┃ precipitation ┃ temp_mean ┃ temp_min ┃ temp_max ┃ year  ┃ month_str ┃ temp_amplitude ┃
┡━━━━━━━━━━━━╇━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━━━━┩
│ date       │ int64 │ string     │ float64    │ float64  │ float64  │ float64          │ float64       │ float64   │ float64  │ float64  │ int32 │ string    │ float64        │
├────────────┼───────┼────────────┼────────────┼──────────┼──────────┼──────────────────┼───────────────┼───────────┼──────────┼──────────┼───────┼───────────┼────────────────┤
│ 2000-01-01 │     1 │ MONTELIMAR │        3.8 │     0.85 │   1.0269 │             0.30 │          0.00 │       5.5 │      2.5 │      8.5 │  2000 │ Jan       │            6.0 │
│ 2000-01-01 │     1 │ PERPIGNAN  │        4.4 │     0.71 │   1.0267 │             0.60 │          0.00 │      12.2 │     10.3 │     14.0 │  2000 │ Jan       │            3.7 │
│ 2000-01-01 │     1 │ TOURS      │        1.6 │     0.97 │   1.0275 │             0.25 │          0.04 │       8.5 │      7.2 │      9.8 │  2000 │ Jan       │            2.6 │
│ 2000-01-02 │     1 │ MONTELIMAR │        5.8 │     0.82 │   1.0287 │             0.54 │          0.00 │       8.3 │      6.8 │      9.8 │  2000 │ Jan       │            3.0 │
│ 2000-01-02 │     1 │ PERPIGNAN  │        2.9 │     0.67 │   1.0278 │             0.96 │          0.00 │       9.8 │      5.1 │     14.6 │  2000 │ Jan       │            9.5 │
│ 2000-01-02 │     1 │ TOURS      │        2.0 │     0.99 │   1.0293 │             0.17 │          0.16 │       7.9 │      6.6 │      9.2 │  2000 │ Jan       │            2.6 │
│ 2000-01-03 │     1 │ MONTELIMAR │        0.4 │     0.92 │   1.0316 │             0.53 │          0.00 │       3.2 │     -2.1 │      8.5 │  2000 │ Jan       │           10.6 │
│ 2000-01-03 │     1 │ PERPIGNAN  │        2.5 │     0.85 │   1.0288 │             0.93 │          0.00 │       8.6 │      4.1 │     13.2 │  2000 │ Jan       │            9.1 │
│ 2000-01-03 │     1 │ TOURS      │        3.4 │     0.91 │   1.0267 │             0.27 │          0.00 │       8.1 │      6.6 │      9.6 │  2000 │ Jan       │            3.0 │
│ 2000-01-04 │     1 │ MONTELIMAR │        1.1 │     0.85 │   1.0274 │             0.64 │          0.00 │       7.2 │      2.3 │     12.1 │  2000 │ Jan       │            9.8 │
│ …          │     … │ …          │          … │        … │        … │                … │             … │         … │        … │        … │     … │ …         │              … │
└────────────┴───────┴────────────┴────────────┴──────────┴──────────┴──────────────────┴───────────────┴───────────┴──────────┴──────────┴───────┴───────────┴────────────────┘

Now that we have used dataset to wrangle our dataset into something we can use for analysis, let's demonstrate how we can use selectors for analysis.

Let's imagine we want to use this dataset for a machine learning project and we need to normalize, calculate the z-score, for all our numeric columns. Selectors make this task relatively straighforward:

In [58]:
(
    weather
    .mutate(
        s.across(
            s.numeric(),
            ( _ - _.mean() / _.std())
        )
    )
)

┏━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┓
┃ date       ┃ month     ┃ city       ┃ wind_speed ┃ humidity  ┃ pressure   ┃ global_radiation ┃ precipitation ┃ temp_mean ┃ temp_min  ┃ temp_max  ┃ year        ┃ month_str ┃ temp_amplitude ┃
┡━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━━━━┩
│ date       │ float64   │ string     │ float64    │ float64   │ float64    │ float64          │ float64       │ float64   │ float64   │ float64   │ float64     │ string    │ float64        │
├────────────┼───────────┼────────────┼────────────┼───────────┼────────────┼──────────────────┼───────────────┼───────────┼───────────┼───────────┼─────────────┼───────────┼────────────────┤
│ 2000-01-01 │ -0.890087 │ MONTELIMAR │   1.979294 │ -4.101485 │ -83.700451 │        -1.325204 │     -0.268422 │  3.447706 │  0.971784 │  6.150888 │ 1302.539633 │ Jan       │       3.734258 │
│ 2000-01-01 │ -0.890087 │ PERPIGNAN  │   2.579294 │ -4.241485 │ -83.700651 │        -1.025204 │     -0.268422 │ 10.147706 │  8.771784 │ 11.650888 │ 1302.539633 │ Jan       │       1.434258 │
│ 2000-01-01 │ -0.890087 │ TOURS      │  -0.220706 │ -3.981485 │ -83.699851 │        -1.375204 │     -0.228422 │  6.447706 │  5.671784 │  7.450888 │ 1302.539633 │ Jan       │       0.334258 │
│ 2000-01-02 │ -0.890087 │ MONTELIMAR │   3.979294 │ -4.131485 │ -83.698651 │        -1.085204 │     -0.268422 │  6.247706 │  5.271784 │  7.450888 │ 1302.539633 │ Jan       │       0.734258 │
│ 2000-01-02 │ -0.890087 │ PERPIGNAN  │   1.079294 │ -4.281485 │ -83.699551 │        -0.665204 │     -0.268422 │  7.747706 │  3.571784 │ 12.250888 │ 1302.539633 │ Jan       │       7.234258 │
│ 2000-01-02 │ -0.890087 │ TOURS      │   0.179294 │ -3.961485 │ -83.698051 │        -1.455204 │     -0.108422 │  5.847706 │  5.071784 │  6.850888 │ 1302.539633 │ Jan       │       0.334258 │
│ 2000-01-03 │ -0.890087 │ MONTELIMAR │  -1.420706 │ -4.031485 │ -83.695751 │        -1.095204 │     -0.268422 │  1.147706 │ -3.628216 │  6.150888 │ 1302.539633 │ Jan       │       8.334258 │
│ 2000-01-03 │ -0.890087 │ PERPIGNAN  │   0.679294 │ -4.101485 │ -83.698551 │        -0.695204 │     -0.268422 │  6.547706 │  2.571784 │ 10.850888 │ 1302.539633 │ Jan       │       6.834258 │
│ 2000-01-03 │ -0.890087 │ TOURS      │   1.579294 │ -4.041485 │ -83.700651 │        -1.355204 │     -0.268422 │  6.047706 │  5.071784 │  7.250888 │ 1302.539633 │ Jan       │       0.734258 │
│ 2000-01-04 │ -0.890087 │ MONTELIMAR │  -0.720706 │ -4.101485 │ -83.699951 │        -0.985204 │     -0.268422 │  5.147706 │  0.771784 │  9.750888 │ 1302.539633 │ Jan       │       7.534258 │
│ …          │         … │ …          │          … │         … │          … │                … │             … │         … │         … │         … │           … │ …         │              … │
└────────────┴───────────┴────────────┴────────────┴───────────┴────────────┴──────────────────┴───────────────┴───────────┴───────────┴───────────┴─────────────┴───────────┴────────────────┘

It seemed we were a little too eager, and we also converted the original month column that was stored as an integer, as well as the year. We can use the `~` selector to omit these columns:

In [59]:
(
    weather
    .mutate(
        s.across(
            (s.numeric() & ~ s.c("month", "year")),
            ( (_ - _.mean()) / _.std())
        )
    )
)

┏━━━━━━━━━━━━┳━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┓
┃ date       ┃ month ┃ city       ┃ wind_speed ┃ humidity  ┃ pressure ┃ global_radiation ┃ precipitation ┃ temp_mean ┃ temp_min  ┃ temp_max  ┃ year  ┃ month_str ┃ temp_amplitude ┃
┡━━━━━━━━━━━━╇━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━━━━┩
│ date       │ int64 │ string     │ float64    │ float64   │ float64  │ float64          │ float64       │ float64   │ float64   │ float64   │ int32 │ string    │ float64        │
├────────────┼───────┼────────────┼────────────┼───────────┼──────────┼──────────────────┼───────────────┼───────────┼───────────┼───────────┼───────┼───────────┼────────────────┤
│ 2000-01-01 │     1 │ MONTELIMAR │  -0.095002 │  0.992571 │ 0.847649 │        -1.315925 │     -0.268422 │ -1.255193 │ -1.133147 │ -1.278583 │  2000 │ Jan       │      -0.752119 │
│ 2000-01-01 │     1 │ PERPIGNAN  │   0.177477 │  0.013550 │ 0.830982 │        -1.006647 │     -0.268422 │ -0.284178 │  0.099470 │ -0.585888 │  2000 │ Jan       │      -1.332341 │
│ 2000-01-01 │     1 │ TOURS      │  -1.094094 │  1.831731 │ 0.897649 │        -1.367472 │     -0.213852 │ -0.820410 │ -0.390416 │ -1.114855 │  2000 │ Jan       │      -1.609838 │
│ 2000-01-02 │     1 │ MONTELIMAR │   0.813263 │  0.782780 │ 0.997649 │        -1.068503 │     -0.268422 │ -0.849395 │ -0.453627 │ -1.114855 │  2000 │ Jan       │      -1.508930 │
│ 2000-01-02 │     1 │ PERPIGNAN  │  -0.503722 │ -0.266171 │ 0.922649 │        -0.635513 │     -0.268422 │ -0.632004 │ -0.722274 │ -0.510321 │  2000 │ Jan       │       0.130828 │
│ 2000-01-02 │     1 │ TOURS      │  -0.912441 │  1.971592 │ 1.047649 │        -1.449946 │     -0.050141 │ -0.907366 │ -0.485233 │ -1.190422 │  2000 │ Jan       │      -1.609838 │
│ 2000-01-03 │     1 │ MONTELIMAR │  -1.639053 │  1.482081 │ 1.239315 │        -1.078812 │     -0.268422 │ -1.588526 │ -1.860075 │ -1.278583 │  2000 │ Jan       │       0.408325 │
│ 2000-01-03 │     1 │ PERPIGNAN  │  -0.685375 │  0.992571 │ 1.005982 │        -0.666441 │     -0.268422 │ -0.805917 │ -0.880302 │ -0.686644 │  2000 │ Jan       │       0.029919 │
│ 2000-01-03 │     1 │ TOURS      │  -0.276655 │  1.412151 │ 0.830982 │        -1.346853 │     -0.268422 │ -0.878381 │ -0.485233 │ -1.140044 │  2000 │ Jan       │      -1.508930 │
│ 2000-01-04 │     1 │ MONTELIMAR │  -1.321160 │  0.992571 │ 0.889315 │        -0.965410 │     -0.268422 │ -1.008816 │ -1.164752 │ -0.825183 │  2000 │ Jan       │       0.206509 │
│ …          │     … │ …          │          … │         … │        … │                … │             … │         … │         … │         … │     … │ …         │              … │
└────────────┴───────┴────────────┴────────────┴───────────┴──────────┴──────────────────┴───────────────┴───────────┴───────────┴───────────┴───────┴───────────┴────────────────┘

We can use selectors within a `group_by`. That is helpful if we wanted to normalize by city and months (the two columns stored as strings):

In [60]:
(
    norm_city_month_weather := weather
    .group_by(s.of_type("string"))
    .mutate(
        s.across(
            (s.numeric() & ~ s.c("date", "month", "year")),
            ( (_ - _.mean()) / _.std())
        )
    )
)

┏━━━━━━━━━━━━┳━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┓
┃ date       ┃ month ┃ city       ┃ wind_speed ┃ humidity  ┃ pressure  ┃ global_radiation ┃ precipitation ┃ temp_mean ┃ temp_min  ┃ temp_max  ┃ year  ┃ month_str ┃ temp_amplitude ┃
┡━━━━━━━━━━━━╇━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━━━━┩
│ date       │ int64 │ string     │ float64    │ float64   │ float64   │ float64          │ float64       │ float64   │ float64   │ float64   │ int32 │ string    │ float64        │
├────────────┼───────┼────────────┼────────────┼───────────┼───────────┼──────────────────┼───────────────┼───────────┼───────────┼───────────┼───────┼───────────┼────────────────┤
│ 2009-11-13 │    11 │ MONTELIMAR │  -1.086555 │  0.841049 │  0.115375 │        -1.455238 │     -0.312727 │ -0.124514 │ -1.026294 │  0.760003 │  2009 │ Nov       │       2.028375 │
│ 2007-11-30 │    11 │ MONTELIMAR │  -0.028354 │ -0.733025 │ -0.184625 │         0.322540 │     -0.312727 │ -0.184376 │ -0.522797 │  0.189612 │  2007 │ Nov       │       0.799119 │
│ 2001-11-18 │    11 │ MONTELIMAR │  -0.353955 │  0.933642 │  0.365375 │        -1.645714 │     -0.210909 │ -0.842861 │ -0.215105 │ -1.262294 │  2001 │ Nov       │      -1.259885 │
│ 2001-11-23 │    11 │ MONTELIMAR │   1.274047 │ -1.103395 │  0.602875 │        -0.598095 │     -0.312727 │ -1.172104 │ -0.970350 │ -1.132659 │  2001 │ Nov       │      -0.276480 │
│ 2001-11-29 │    11 │ MONTELIMAR │   0.582146 │ -0.547840 │  0.902875 │         0.449524 │     -0.123636 │ -0.363963 │ -0.159161 │ -0.458560 │  2001 │ Nov       │      -0.368674 │
│ 2001-11-30 │    11 │ MONTELIMAR │   0.093746 │  0.470679 │  0.240375 │        -0.026667 │     -0.312727 │  0.204729 │  0.064615 │  0.293320 │  2001 │ Nov       │       0.276685 │
│ 2002-11-03 │    11 │ MONTELIMAR │  -1.045855 │  0.378086 │ -0.184625 │        -0.185397 │     -0.283636 │  1.491769 │  1.267413 │  1.382249 │  2002 │ Nov       │       0.245954 │
│ 2002-11-05 │    11 │ MONTELIMAR │   0.826346 │ -0.640432 │  0.640375 │         0.893968 │     -0.312727 │  0.623765 │  0.903776 │  0.241466 │  2002 │ Nov       │      -0.706720 │
│ 2002-11-11 │    11 │ MONTELIMAR │  -0.842355 │  0.655864 │ -0.372125 │        -0.693333 │     -0.312727 │  0.863215 │  0.540140 │  0.993345 │  2002 │ Nov       │       0.583999 │
│ 2002-11-12 │    11 │ MONTELIMAR │  -1.127255 │  0.192901 │ -0.322125 │         1.116190 │      0.050909 │ -0.004789 │ -0.914406 │  0.837784 │  2002 │ Nov       │       1.997644 │
│ …          │     … │ …          │          … │         … │         … │                … │             … │         … │         … │         … │     … │ …         │              … │
└────────────┴───────┴────────────┴────────────┴───────────┴───────────┴──────────────────┴───────────────┴───────────┴───────────┴───────────┴───────┴───────────┴────────────────┘

The `s.across` function also allows you to do multiple operations on your columns at once. For instance, if we also wanted to center the values, we can specify a dict. The names will be appended at the end of the columns:

In [61]:
(
     weather
    .group_by(s.of_type("string"))
    .mutate(
        s.across(
            (s.numeric() & ~ s.c("date", "month", "year")),
            dict(
                centered = (_ - _.min()) / (_.max() - _.min()),
                zscore = (_ - _.mean()) / _.std()
            )
        )
    )
    .select(
        s.of_type("string"),
        s.endswith(("_centered", "_zscore"))
    )
)

┏━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━┓
┃ city       ┃ month_str ┃ wind_speed_centered ┃ humidity_centered ┃ pressure_centered ┃ global_radiation_centered ┃ precipitation_centered ┃ temp_mean_centered ┃ temp_min_centered ┃ temp_max_centered ┃ temp_amplitude_centered ┃ wind_speed_zscore ┃ humidity_zscore ┃ pressure_zscore ┃ global_radiation_zscore ┃ precipitation_zscore ┃ temp_mean_zscore ┃ temp_min_zscore ┃ temp_max_zscore ┃ temp_amplitude_zscore ┃
┡━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━┩
│ string     │ string    │ float64             │ float64           │ float64           │ float64                   │ float64                │ float64            │ float64           │ float64           │ float64                 │ float64           │ float64         │ float64         │ float64                 │ float64              │ float64          │ float64         │ float64         │ float64               │
├────────────┼───────────┼─────────────────────┼───────────────────┼───────────────────┼───────────────────────────┼────────────────────────┼────────────────────┼───────────────────┼───────────────────┼─────────────────────────┼───────────────────┼─────────────────┼─────────────────┼─────────────────────────┼──────────────────────┼──────────────────┼─────────────────┼─────────────────┼───────────────────────┤
│ MONTELIMAR │ Nov       │            0.632075 │          0.392157 │          0.737209 │                  0.365079 │               0.000000 │           0.320225 │          0.339901 │          0.289617 │                0.275641 │          1.274047 │       -1.103395 │        0.602875 │               -0.598095 │            -0.312727 │        -1.172104 │       -0.970350 │       -1.132659 │             -0.276480 │
│ MONTELIMAR │ Nov       │            0.122642 │          0.647059 │          0.769767 │                  0.722222 │               0.000000 │           0.685393 │          0.497537 │          0.819672 │                0.692308 │         -0.923755 │        0.100309 │        0.777875 │                0.830476 │            -0.312727 │         0.773421 │       -0.075245 │        1.382249 │              1.721061 │
│ MONTELIMAR │ Nov       │            0.056604 │          0.862745 │          0.590698 │                  0.619048 │               0.000000 │           0.438202 │          0.305419 │          0.551913 │                0.628205 │         -1.208656 │        1.118827 │       -0.184625 │                0.417778 │            -0.312727 │        -0.543550 │       -1.166154 │        0.111831 │              1.413747 │
│ MONTELIMAR │ Nov       │            0.301887 │          0.686275 │          0.000000 │                  0.500000 │               0.020344 │           0.724719 │          0.724138 │          0.650273 │                0.198718 │         -0.150454 │        0.285494 │       -3.359625 │               -0.058413 │            -0.123636 │         0.982939 │        1.211469 │        0.578515 │             -0.645257 │
│ MONTELIMAR │ Nov       │            0.160377 │          0.705882 │          0.339535 │                  0.444444 │               0.004695 │           0.460674 │          0.423645 │          0.475410 │                0.384615 │         -0.760955 │        0.378086 │       -1.534625 

We can even use the `names` argument to specify how the new columns are going to be named, either using lambda functions or format strings:

In [62]:
(
     weather
    .group_by(s.of_type("string"))
    .mutate(
        s.across(
            (s.numeric() & ~ s.c("date", "month", "year")),
            dict(
                centered = (_ - _.min()) / (_.max() - _.min()),
                zscore = (_ - _.mean()) / _.std()
            ),
            names = lambda col, fn: f"new_{fn}_{col}"
        )
    )
    .select(
        s.of_type("string"),
        s.startswith("new_")
    )
)

┏━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ city       ┃ month_str ┃ new_centered_wind_speed ┃ new_centered_humidity ┃ new_centered_pressure ┃ new_centered_global_radiation ┃ new_centered_precipitation ┃ new_centered_temp_mean ┃ new_centered_temp_min ┃ new_centered_temp_max ┃ new_centered_temp_amplitude ┃ new_zscore_wind_speed ┃ new_zscore_humidity ┃ new_zscore_pressure ┃ new_zscore_global_radiation ┃ new_zscore_precipitation ┃ new_zscore_temp_mean ┃ new_zscore_temp_min ┃ new_zscore_temp_max ┃ new_zscore_temp_amplitude ┃
┡━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ string     │ string    │ float64                 │ float64               │ float64               │ float64                       │ float64                    │ float64                │ float64               │ float64               │ float64                     │ float64               │ float64             │ float64             │ float64                     │ float64                  │ float64              │ float64             │ float64             │ float64                   │
├────────────┼───────────┼─────────────────────────┼───────────────────────┼───────────────────────┼───────────────────────────────┼────────────────────────────┼────────────────────────┼───────────────────────┼───────────────────────┼─────────────────────────────┼───────────────────────┼─────────────────────┼─────────────────────┼─────────────────────────────┼──────────────────────────┼──────────────────────┼─────────────────────┼─────────────────────┼───────────────────────────┤
│ MONTELIMAR │ Nov       │                0.084906 │              0.803922 │              0.646512 │                      0.150794 │                   0.000000 │               0.516854 │              0.330049 │              0.688525 │                    0.756410 │             -1.086555 │            0.841049 │            0.115375 │                   -1.455238 │                -0.312727 │            -0.124514 │           -1.026294 │            0.760003 │                  2.028375 │
│ MONTELIMAR │ Nov       │                0.075472 │              0.882353 │              0.644186 │                      0.261905 │                   0.040689 │               0.617978 │              0.502463 │              0.683060 │                    0.525641 │             -1.127255 │            1.211420 │            0.102875 │                   -1.010794 │                 0.065455 │             0.414247 │           -0.047273 │            0.734077 │                  0.922045 │
│ MONTELIMAR │ Nov       │                0.254717 │              0.823529 │              0.693023 │                      0.103175 │                   0.010955 │               0.382022 │              0.472906 │              0.262295 │                    0.070513 │             -0.353955 │            0.933642 │            0.365375 │                   -1.645714 │                -0.210909 │            -0.842861 │           -0.215105 │           -1.262294 │                 -1.259885 │
│ MONTELIMAR │ Nov       │                0.632075 │              0.392157 │              0.737209 │                    

In [63]:
(
     weather
    .group_by(s.of_type("string"))
    .mutate(
        s.across(
            (s.numeric() & ~ s.c("date", "month", "year")),
            dict(
                centered = (_ - _.min()) / (_.max() - _.min()),
                zscore = (_ - _.mean()) / _.std()
            ),
            names = "new_{fn}_{col}"
        )
    )
    .select(
        s.of_type("string"),
        s.startswith("new_")
    )
)

┏━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃ city       ┃ month_str ┃ new_centered_wind_speed ┃ new_centered_humidity ┃ new_centered_pressure ┃ new_centered_global_radiation ┃ new_centered_precipitation ┃ new_centered_temp_mean ┃ new_centered_temp_min ┃ new_centered_temp_max ┃ new_centered_temp_amplitude ┃ new_zscore_wind_speed ┃ new_zscore_humidity ┃ new_zscore_pressure ┃ new_zscore_global_radiation ┃ new_zscore_precipitation ┃ new_zscore_temp_mean ┃ new_zscore_temp_min ┃ new_zscore_temp_max ┃ new_zscore_temp_amplitude ┃
┡━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━┩
│ string     │ string    │ float64                 │ float64               │ float64               │ float64                       │ float64                    │ float64                │ float64               │ float64               │ float64                     │ float64               │ float64             │ float64             │ float64                     │ float64                  │ float64              │ float64             │ float64             │ float64                   │
├────────────┼───────────┼─────────────────────────┼───────────────────────┼───────────────────────┼───────────────────────────────┼────────────────────────────┼────────────────────────┼───────────────────────┼───────────────────────┼─────────────────────────────┼───────────────────────┼─────────────────────┼─────────────────────┼─────────────────────────────┼──────────────────────────┼──────────────────────┼─────────────────────┼─────────────────────┼───────────────────────────┤
│ MONTELIMAR │ Nov       │                0.773585 │              0.372549 │              0.700000 │                      0.642857 │                   0.000000 │               0.595506 │              0.635468 │              0.502732 │                    0.141026 │              1.884548 │           -1.195988 │            0.402875 │                    0.513016 │                -0.312727 │             0.294523 │            0.707972 │           -0.121511 │                 -0.921840 │
│ MONTELIMAR │ Nov       │                0.122642 │              0.647059 │              0.769767 │                      0.722222 │                   0.000000 │               0.685393 │              0.497537 │              0.819672 │                    0.692308 │             -0.923755 │            0.100309 │            0.777875 │                    0.830476 │                -0.312727 │             0.773421 │           -0.075245 │            1.382249 │                  1.721061 │
│ MONTELIMAR │ Nov       │                0.056604 │              0.862745 │              0.590698 │                      0.619048 │                   0.000000 │               0.438202 │              0.305419 │              0.551913 │                    0.628205 │             -1.208656 │            1.118827 │           -0.184625 │                    0.417778 │                -0.312727 │            -0.543550 │           -1.166154 │            0.111831 │                  1.413747 │
│ MONTELIMAR │ Nov       │                0.301887 │              0.686275 │              0.000000 │                    

We can also use selectors with aggregations. For instance, if we want to calculate the average temperatures for the minimum, maximum, and mean, per city and month:

In [64]:
(
    weather
    .group_by(s.of_type("string"))
    .agg(
        s.across(
            s.startswith("temp"),
            _.mean()
        )
    )
    .filter(_.city == "PERPIGNAN")
)

┏━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┓
┃ city      ┃ month_str ┃ temp_mean ┃ temp_min  ┃ temp_max  ┃ temp_amplitude ┃
┡━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━━━━┩
│ string    │ string    │ float64   │ float64   │ float64   │ float64        │
├───────────┼───────────┼───────────┼───────────┼───────────┼────────────────┤
│ PERPIGNAN │ Jan       │  8.480064 │  4.331833 │ 12.621865 │       8.290032 │
│ PERPIGNAN │ Feb       │  9.259011 │  5.064664 │ 13.448410 │       8.383746 │
│ PERPIGNAN │ Mar       │ 11.815161 │  7.423226 │ 16.210323 │       8.787097 │
│ PERPIGNAN │ Apr       │ 14.219000 │  9.892667 │ 18.547000 │       8.654333 │
│ PERPIGNAN │ May       │ 17.991935 │ 13.470000 │ 22.514839 │       9.044839 │
│ PERPIGNAN │ Jun       │ 22.571667 │ 17.716333 │ 27.429000 │       9.712667 │
│ PERPIGNAN │ Jul       │ 24.447419 │ 19.559355 │ 29.334516 │       9.775161 │
│ PERPIGNAN │ Aug       │ 24.438065 │ 19.537097 │ 29.345161 │       9.808065 │
│ PERPIGNAN │ Sep       │ 20.645333 │ 16.021333 │ 25.271667 │       9.250333 │
│ PERPIGNAN │ Oct       │ 17.294194 │ 13.139032 │ 21.447419 │       8.308387 │
│ …         │ …         │         … │         … │         … │              … │
└───────────┴───────────┴───────────┴───────────┴───────────┴────────────────┘

To finish, let's say we want to identify the extreme temperature events in our dataset. We can calculate the z-scores grouped by city and month, and filter all the days that have an absolute z-score greater than 2 in any of the temperature measure columns:

In [68]:
(
    weather
    .group_by((s.of_type("string")))
    .mutate(
        s.across(
            s.startswith("temp"),
            dict(zscore = (_ - _.mean()) / _.std())
        )
    )
    .filter(
        s.if_any(
            s.startswith("temp") & s.endswith("_zscore"),
            _.abs() > 2
        )
    )
    .select(
        s.c("date"),
        s.of_type("string"),
        s.startswith("temp")
    )
)

┏━━━━━━━━━━━━┳━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━┓
┃ date       ┃ city       ┃ month_str ┃ temp_mean ┃ temp_min ┃ temp_max ┃ temp_amplitude ┃ temp_mean_zscore ┃ temp_min_zscore ┃ temp_max_zscore ┃ temp_amplitude_zscore ┃
┡━━━━━━━━━━━━╇━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━┩
│ date       │ string     │ string    │ float64   │ float64  │ float64  │ float64        │ float64          │ float64         │ float64         │ float64               │
├────────────┼────────────┼───────────┼───────────┼──────────┼──────────┼────────────────┼──────────────────┼─────────────────┼─────────────────┼───────────────────────┤
│ 2000-11-13 │ MONTELIMAR │ Nov       │      16.6 │     15.2 │     18.1 │            2.9 │         2.090392 │        2.498182 │        1.330395 │             -1.167691 │
│ 2003-11-23 │ MONTELIMAR │ Nov       │      17.7 │     15.1 │     20.2 │            5.1 │         2.419635 │        2.470210 │        1.874860 │             -0.491600 │
│ 2005-11-03 │ MONTELIMAR │ Nov       │      17.4 │     13.1 │     21.6 │            8.5 │         2.329841 │        1.910769 │        2.237836 │              0.553268 │
│ 2008-11-01 │ MONTELIMAR │ Nov       │      16.4 │     13.7 │     19.2 │            5.5 │         2.030530 │        2.078601 │        1.615591 │             -0.368674 │
│ 2008-11-28 │ MONTELIMAR │ Nov       │       2.9 │      0.9 │      4.8 │            3.9 │        -2.010177 │       -1.501818 │       -2.117881 │             -0.860377 │
│ 2005-11-21 │ MONTELIMAR │ Nov       │       2.7 │     -0.4 │      5.8 │            6.2 │        -2.070039 │       -1.865455 │       -1.858612 │             -0.153555 │
│ 2005-11-24 │ MONTELIMAR │ Nov       │       1.2 │     -0.8 │      3.3 │            4.1 │        -2.519006 │       -1.977343 │       -2.506784 │             -0.798914 │
│ 2005-11-25 │ MONTELIMAR │ Nov       │       2.6 │      0.5 │      4.6 │            4.1 │        -2.099970 │       -1.613706 │       -2.169735 │             -0.798914 │
│ 2005-11-27 │ MONTELIMAR │ Nov       │       0.8 │     -1.9 │      3.5 │            5.4 │        -2.638731 │       -2.285035 │       -2.454930 │             -0.399406 │
│ 2005-11-28 │ MONTELIMAR │ Nov       │       0.0 │     -4.1 │      4.2 │            8.3 │        -2.878180 │       -2.900420 │       -2.273442 │              0.491805 │
│ …          │ …          │ …         │         … │        … │        … │              … │                … │               … │               … │                     … │
└────────────┴────────────┴───────────┴───────────┴──────────┴──────────┴────────────────┴──────────────────┴─────────────────┴─────────────────┴───────────────────────┘

You probably wouldn't want to write the equivalent SQL query:

In [69]:
ibis.show_sql((
     weather
    .group_by((s.of_type("string")))
    .mutate(
        s.across(
            s.startswith("temp"),
            dict(zscore = (_ - _.mean()) / _.std())
        )
    )
    .filter(
        s.if_any(
            s.startswith("temp") & s.endswith("_zscore"),
            _.abs() > 2
        )
    )
    .select(
        s.c("date"),
        s.of_type("string"),
        s.startswith("temp")
    )
))

WITH t0 AS (
  SELECT
    t9."DATE" AS "DATE",
    t9."MONTH" AS "MONTH",
    t9."MONTELIMAR_wind_speed" AS "MONTELIMAR_wind_speed",
    t9."MONTELIMAR_humidity" AS "MONTELIMAR_humidity",
    t9."MONTELIMAR_pressure" AS "MONTELIMAR_pressure",
    t9."MONTELIMAR_global_radiation" AS "MONTELIMAR_global_radiation",
    t9."MONTELIMAR_precipitation" AS "MONTELIMAR_precipitation",
    t9."MONTELIMAR_temp_mean" AS "MONTELIMAR_temp_mean",
    t9."MONTELIMAR_temp_min" AS "MONTELIMAR_temp_min",
    t9."MONTELIMAR_temp_max" AS "MONTELIMAR_temp_max",
    t9."PERPIGNAN_wind_speed" AS "PERPIGNAN_wind_speed",
    t9."PERPIGNAN_humidity" AS "PERPIGNAN_humidity",
    t9."PERPIGNAN_pressure" AS "PERPIGNAN_pressure",
    t9."PERPIGNAN_global_radiation" AS "PERPIGNAN_global_radiation",
    t9."PERPIGNAN_precipitation" AS "PERPIGNAN_precipitation",
    t9."PERPIGNAN_temp_mean" AS "PERPIGNAN_temp_mean",
    t9."PERPIGNAN_temp_min" AS "PERPIGNAN_temp_min",
    t9."PERPIGNAN_temp_max" AS "PERPIGNAN_temp_max"